jobs/cloud/worker.go

```diff
@@ -0,0 +1,284 @@
+		subscription, subErr = worker.app.Cloud().GetSubscription()
+
+		if subErr != nil {
+			mlog.Error("Worker: Failed to retrieve subscription", mlog.String("worker", worker.name), mlog.String("job_id", job.Id), mlog.String("error", subErr.Error()))
+			worker.setJobError(job, subErr)
+			return
+		}
+
+		if subscription == nil {
+			mlog.Error("Worker: Failed to retrieve subscription", mlog.String("worker", worker.name), mlog.String("job_id", job.Id))
+			return
+		}
+	}
+
+	if subscription != nil && subscription.IsPaidTier == "true" {
+		mlog.Info("On Paid Tier, exiting", mlog.String("worker", worker.name))
+		return
+	}
+
+	count, err := worker.app.Srv().Store.User().Count(model.UserCountOptions{IncludeDeleted: false})
+
+	if err != nil {
+		mlog.Error("Worker: Failed to get active user count", mlog.String("worker", worker.name), mlog.String("job_id", job.Id), mlog.String("error", err.Error()))
+		return
+	}
+
+	userDifference := cloudUserLimit - count
+
+	systemValue, nErr := worker.app.Srv().Store.System().GetByName(model.USER_LIMIT_OVERAGE_CYCLE_END_DATE)
+	dateLayout := "2006-01-02"
+
+	if nErr != nil {
+		mlog.Error("Error getting USER_LIMIT_OVERAGE_CYCLE_END_DATE from system store", mlog.String("worker", worker.name), mlog.String("error", nErr.Error()))
+	}
+
+	if systemValue == nil {
+		mlog.Info("Found nil end date")
+		if userDifference >= 0 {
+			// Under user limit, so no need to start tracking yet.
+			worker.LogAndSetJobSuccess(job)
+			return
+		}
+		// Our row doesn't exist, so create it and get it from the store
+		subCycleEndDate := time.Unix((subscription.EndAt / 1000), 0)
+		sysVar := &model.System{Name: model.USER_LIMIT_OVERAGE_CYCLE_END_DATE, Value: subCycleEndDate.Format(dateLayout)}
+		err := worker.app.Srv().Store.System().SaveOrUpdate(sysVar)
+		if err != nil {
+			mlog.Error("Unable to save USER_LIMIT_OVERAGE_CYCLE_END_DATE count", mlog.String("worker", worker.name), mlog.String("error", nErr.Error()))
+		}
+
+		// Store today's date (first day over limit) so we can reference it later
+		t := time.Now()
+		firstDayOverVar := &model.System{Name: model.OVER_USER_LIMIT_DATE, Value: t.Format(dateLayout)}
+		err = worker.app.Srv().Store.System().SaveOrUpdate(firstDayOverVar)
+		if err != nil {
+			mlog.Error("Unable to save USER_LIMIT_OVERAGE_CYCLE_END_DATE count", mlog.String("worker", worker.name), mlog.String("error", nErr.Error()))
+		}
+		worker.LogAndSetJobSuccess(job)
+		return
+	}
+
+	// If we've reached this point, we know at some time the installation was over the limit
+
+	subCycleEndDate, err := time.Parse(dateLayout, systemValue.Value)
+	if err != nil {
+		mlog.Error("Unable to parse USER_LIMIT_OVERAGE_CYCLE_END_DATE", mlog.String("worker", worker.name), mlog.String("error", err.Error()))
+	}
+
+	now := time.Now()
+
+	daysDifference := int64(math.Floor(now.Sub(subCycleEndDate).Hours() / 24)) // Get the difference in days between now and cycle end date
```

Comment:
We need to send the email 7/14/30 days after the billing cycle right? So here wouldn't it be an addition?

So if the billing cycle end date is Oct. 31st., the first email to send would be Nov. 7th?

Reply:
At this point using your example, the `subCycleEndDate` would be October 31st. When we calculate the difference between the time.Now(), and the subCycleEndDate, we get the difference in days between the subCycleEndDate and time.Now(). If time.Now() is any date after Oct 31, we will have a positive number that can be used in the switch statement (when that number is 7, 14, 30, etc, it triggers a different case). If it's negative, it won't trigger a case, because that means we're still in the current cycle.


I could have done this by saving the next date of the check but figured it would be a bit cleaner to just calculate the difference each time.

Reply:
Gotcha! Thanks for the explanation!